core/shared/src/main/scala/zio/Cause.scala

```diff
@@ -754,6 +755,59 @@ sealed abstract class Cause[+E] extends Product with Serializable { self =>
     e.addSuppressed(trace)
     e
   }
+
+  /**
+   * Stack-safe toString for Cause
+   */
+  final private def causeToString: String = {
+    // result modifier function or cause to visit
+    val visitStack = new mutable.Stack[Either[() => Unit, Cause[E]]]()
+    val results    = new mutable.Stack[String]()
+
+    def twoArgStr(name: String) = {
+      val right = results.pop()
+      val left  = results.pop()
+      results.push(s"$name($left,$right)")
+    }
+
+    @tailrec
+    def visitRecursive(): String = {
+      def visitCause(current: Cause[E]) =
+        current match {
+          case Both(left, right) =>
+            visitStack.push(Left(() => twoArgStr("Both")), Right(right), Right(left))
+          case Then(left, right) =>
+            visitStack.push(Left(() => twoArgStr("Then")), Right(right), Right(left))
+          case Stackless(cause, stackless) =>
+            visitStack.push(Left(() => results.push(s"Stackless(${results.pop()},$stackless)")), Right(cause))
+          case Die(value, trace) =>
+            results.push(s"Die($value,$trace)")
+          case Empty =>
```

Comment:
Oh I missed this previously. Avoid using `case Empty` because it uses the `equals` method of Cause which can be expensive. This is much more performant as it relies on type-checking alone:

```suggestion
          case _: Empty.type =>
```

Reply:
Remade all checks